File: src/c99/efermi.c
#include "efermi.h"

double efermi(size_t nkpt, size_t nbnd,
              double bands[nkpt][nbnd], double weights[nkpt],
              int nelec,
              double swidth, int stype)
// calculate Fermi energy using bisection
{
    double eigmin, eigmax, x0, x1, x2, dx, xmid, f, fmid, rtb;
    int i, j, k, n;

    // Get min, max eigenvalue and set as initial bounds
    x1 = bands[0][0];
    x2 = bands[0][0];
    for (register int i = 0; i < nkpt; i++)
    {
        for (register int j = 0; j < nbnd; ++j)
        {
            if (bands[i][j] < x1)
                x1 = bands[i][j];
            if (bands[i][j] > x2)
                x2 = bands[i][j];
        }
    }
    x0 = (x1 + x2) * 0.5;

    // Calculate initial f, fmid
    f = smear(nkpt, nbnd, bands, weights, x1, nelec, swidth, stype);
    fmid = smear(nkpt, nbnd, bands, weights, x2, nelec, swidth, stype);

    // Find bounds which bracket the Fermi energy
    for (register int n; n < NMAX; ++n)
    {
        if (f * fmid >= 0.0)
        {
            x1 = x0 - (double)n * swidth;
            x2 = x0 + ((double)n - 0.5) * swidth;
            f = smear(nkpt, nbnd, bands, weights, x1, nelec, swidth, stype);
            fmid = smear(nkpt, nbnd, bands, weights, x2, nelec, swidth, stype);
        }
        else
        {
            break;
        }
    }
    if (f * fmid >= 0.0)
    {
        printf("Could not bracket Fermi energy. Smearing too small?\n");
        return 0.0;
    }

    // Set initial Fermi energy guess
    rtb = f < 0.0 ? (dx = x2 - x1, x1) : (dx = x1 - x2, x2);

    // Do bisection
    for (register int j = 0; j < JMAX; ++j)
    {
        if ((fabs(dx) <= XACC) || (fmid == 0.0))
        {
            return rtb;
        }
        dx = dx * 0.5;
        xmid = rtb + dx;
        fmid = smear(nkpt, nbnd, bands, weights, xmid, nelec, swidth, stype);
        if (fmid <= 0.0)
        {
            rtb = xmid;
        }
    }
    printf("Reached maximum number of bisections.\n");
    return rtb;
}

double smear(size_t nkpt, size_t nbnd,
             double bands[nkpt][nbnd], double weights[nkpt],
             double xe,
             int nelec, float swidth, int stype)
// calculate smeared value used for bisection
{
    double x, z;
    SMEARING_FUNC *smearing_funcs[6] = {gaussian, fermid, delthm, spline, poshm, poshm2};

    z = 0.0;
    for (register int i = 0; i < nkpt; ++i)
    {
        for (register int j = 0; j < nbnd; ++j)
        {
            x = (xe - bands[i][j]) / swidth;
            z = z + weights[i] * smearing_funcs[stype - 1](x);
        }
    }
    return z - (double)nelec;
}

// Gaussian
static inline double gaussian(double x)
{
    return 2.0 - erfc(x);
}

// Fermi-Dirac
static inline double fermid(double x)
{
    // AZ: flip around x-axis so all smearing func.s called the same way
    x = -x;
    if (x > FDCUT)
        return 0.0;
    else if (x < -FDCUT)
        return 2.0;
    else
        return 2.0 / (1.0 + exp(x));
}

// Hermite delta expansion
static inline double delthm(double x)
{
    if (x > HMCUT)
        return 2.0;
    else if (x < -HMCUT)
        return 0.0;
    else
        return (2.0 - erfc(x)) + x * exp(-x * x) * ISQRTPI;
}

// Gaussian spline
static inline double spline(double x)
{
    // AZ: flip around x-axis so all smearing func.s called the same way
    x = -x;
    if (x > 0.0)
        return 2.0 * (HSQRTE * exp(-(x + ISQRT2) * (x + ISQRT2)));
    else
        return 2.0 * (1.0 - HSQRTE * exp(-(x - ISQRT2) * (x - ISQRT2)));
}

// Positive Hermite (cold I)
static inline double poshm(double x)
{
    // NM: NOTE: g's are all intended to be normalized to 1!
    // NM: function = 2 * int_minf^x [g(t)] dt
    if (x > HMCUT)
        return 2.0;
    else if (x < -HMCUT)
        return 0.0;
    else
        return (2.0 - erfc(x)) + (-2.0 * POSHMA * x * x + 2.0 * x + POSHMA) * exp(-x * x) * I2SQRTPI;
}

// Positive Hermite (cold II)
static inline double poshm2(double x)
{
    // NM: NOTE: g's are all intended to be normalized to 1!
    // NM: function = 2 * int_minf^x [g(t)] dt
    if (x > HMCUT)
        return 2.0;
    else if (x < -HMCUT)
        return 0.0;
    else
        return (2.0 - erfc(x - ISQRT2)) + SQRT2 * exp(-x * x + SQRT2 * x - 0.5) * ISQRTPI;
}

File: src/py/efermi.py
import numpy as np

XACC = 1e-6
JMAX = 50
NMAX = 50
FDCUT = 30
HMCUT = 10
POSHMA = -0.5634

PI = np.pi
EE = np.e
SQRT2 = np.sqrt(2)
SQRTPI = np.sqrt(np.pi)
EESH = np.sqrt(np.e) / 2
SQ2I = np.sqrt(2) / 2
PIEESQQ = np.sqrt(np.e * np.pi) / 4


def efermi(bands, weights, nelecs, smearing_width, smearing_type):
    eigmin = np.min(bands)
    eigmax = np.max(bands)

    # print("n xe1      xe2      f          fmid       f * fmid")

    xe1 = eigmin
    xe2 = eigmax
    xe0 = (xe1 + xe2) / 2
    n = 1

    f = smear(bands, weights, xe1, nelecs, smearing_width, smearing_type)
    fmid = smear(bands, weights, xe2, nelecs, smearing_width, smearing_type)

    # print(f"{n} {xe1:0.6f} {xe2:0.6f} {f:0.6f} {fmid:0.6f} {f * fmid:0.6f}")

    while f * fmid >= 0:
        if n >= NMAX:
            raise Exception(
                "Could not bracket Fermi energy. Is the electronic temperature too small?"
            )

        n += 1
        xe1 = xe0 - n * smearing_width
        xe2 = xe0 + (n - 0.5) * smearing_width

        f = smear(bands, weights, xe1, nelecs, smearing_width, smearing_type)
        fmid = smear(bands, weights, xe2, nelecs, smearing_width,
                     smearing_type)

        # print(
        #     f"{n} {xe1:0.6f} {xe2:0.6f} {f:0.6f} {fmid:0.6f} {f * fmid:0.6f}")

    if f < 0:
        rtbis = xe1
        dx = xe2 - xe1
    else:
        rtbis = xe2
        dx = xe1 - xe2

    j = 0
    while (np.abs(dx) > XACC) and (fmid != 0):
        if j >= JMAX:
            raise Exception("Reached maximum number of bisections.")

        dx *= 0.5
        xmid = rtbis + dx

        fmid = smear(bands, weights, xmid, nelecs, smearing_width,
                     smearing_type)

        if fmid <= 0:
            rtbis = xmid

        j += 1

    ef = rtbis

    return ef


def smear(bands, weights, xe, nelecs, smearing_width, smearing_type):
    z = 0
    for i, kpt in enumerate(bands):
        for j, eigval in enumerate(kpt):
            x = (xe - eigval) / smearing_width
            if smearing_type == 1:
                z += weights[i] * (2 - np.erfc(x))
            elif smearing_type == 2:
                z += weights[i] * fermid(-x)
            elif smearing_type == 3:
                z += weights[i] * delthm(x)
            elif smearing_type == 4:
                z += weights[i] * spline(-x)
            elif smearing_type == 5:
                z += weights[i] * poshm(x)
            elif smearing_type == 6:
                z += weights[i] * poshm2(x)
            else:
                raise Exception(f"Unknown smearing type `{smearing_type}`.")
    return z - nelecs


def fermid(xx):
    if xx > FDCUT:
        return 0
    elif xx < -FDCUT:
        return 2
    else:
        return 2 / (1 + np.exp(xx))


def delthm(xx):
    if xx > HMCUT:
        return 2
    elif xx < -HMCUT:
        return 0
    else:
        return (2 - np.erfc(xx)) + xx * np.exp(-xx * xx) / SQRTPI


def spline(x):
    if x > 0:
        return 2 * (EESH * np.exp(-(x + SQ2I) * -(x + SQ2I)))
    else:
        return 2 * (1.0 - EESH * np.exp(-(x + SQ2I) * -(x + SQ2I)))


def poshm(x):
    if x > HMCUT:
        return 2
    elif x < -HMCUT:
        return 0
    else:
        return (2 - np.erfc(x)) + (-2 * POSHMA * x * x + 2 * x +
                                   POSHMA) * np.exp(-x * x) / SQRTPI / 2


def poshm2(x):
    if x > HMCUT:
        return 2
    elif x < -HMCUT:
        return 0
    else:
        return (2 - np.erfc(x - 1 / SQRT2)
                ) + SQRT2 * np.exp(-x * x + SQRT2 * x - 0.5) / SQRTPI
